File: Arbitration/MPArbitration/Model/TDIRequestDetails.cs
using System.Text.Json.Serialization;
using System.ComponentModel.DataAnnotations.Schema;
using Microsoft.EntityFrameworkCore;
using System.ComponentModel.DataAnnotations;

namespace MPArbitration.Model
{

    [Index(nameof(RequestId), nameof(BatchUploadDate))]
    public class TDIRequestDetails
    {
        [JsonPropertyName("id")]
        public int Id { get; set; } = 0;

        [JsonPropertyName("action")]
        [StringLength(20)]
        public string Action { get; set; } = "";

        [JsonPropertyName("additionalPaidAmount")]
        public double AdditionalPaidAmount { get; set; } = 0;

        [JsonPropertyName("arbitrationDate")]
        public DateTime? ArbitrationDate { get; set; } = null;

        [JsonPropertyName("arbitrationDecisionDate")]
        public DateTime? ArbitrationDecisionDate { get; set; } = null;

        [JsonPropertyName("arbitratorReportSubmissionDate")]
        public DateTime? ArbitratorReportSubmissionDate { get; set; } = null;

        [JsonPropertyName("arbitrator1")]
        [StringLength(100)]
        public string Arbitrator1 { get; set; } = "";

        [JsonPropertyName("arbitrator2")]
        [StringLength(100)]
        public string Arbitrator2 { get; set; } = "";

        [JsonPropertyName("arbitrator3")]
        [StringLength(100)]
        public string Arbitrator3 { get; set; } = "";

        [JsonPropertyName("arbitrator4")]
        [StringLength(100)]
        public string Arbitrator4 { get; set; } = "";

        [JsonPropertyName("arbitrator5")]
        [StringLength(100)]
        public string Arbitrator5 { get; set; } = "";

        [JsonPropertyName("assignmentDeadlineDate")]
        public DateTime? AssignmentDeadlineDate { get; set; } = null;

        [JsonPropertyName("batchUploadDate")]
        public DateTime BatchUploadDate { get; set; } = new DateTime();

        [JsonPropertyName("payorClaimNumber")]
        [StringLength(50)]
        public string PayorClaimNumber { get; set; } = "";

        [JsonPropertyName("createdBy")]
        [StringLength(60)]
        public string CreatedBy { get; set; } = "";

        [JsonPropertyName("daysOpenCount")]
        public int DaysOpenCount { get; set; } = 0;

        [JsonPropertyName("disputedAmount")]
        public double DisputedAmount { get; set; } = 0;

        [NotMapped]
        [JsonPropertyName("DOB")]
        public DateTime? DOB { get; set; } = null;

        [JsonPropertyName("entity")]
        [StringLength(60)]
        public string Entity { get; set; } = ""; /* provider entity */

        [JsonPropertyName("entityNPI")]
        [StringLength(10)]
        public string EntityNPI { get; set; } = "";

        [JsonPropertyName("estimatedDisputedAmount")]
        public double EstimatedDisputedAmount { get; set; } = 0;

        [JsonPropertyName("healthPlanName")]
        [StringLength(100)]
        public string HealthPlanName { get; set; } = ""; /* Payor */

        [JsonPropertyName("history")]
        [StringLength(255)]
        public string History { get; set; } = "";

        [JsonPropertyName("ineligibilityReason")]
        [StringLength(2048)]
        public string IneligibilityReason { get; set; } = "";

        [JsonPropertyName("informalTeleconferenceDate")]
        public DateTime? InformalTeleconferenceDate { get; set; } = null;

        [JsonPropertyName("methodOfPayment")]
        [StringLength(50)]
        public string MethodOfPayment { get; set; } = "";

        [JsonPropertyName("originalBilledAmount")]
        public double OriginalBilledAmount { get; set; } = 0;

        [JsonPropertyName("partiesAwardNotificationDate")]
        public DateTime? PartiesAwardNotificationDate { get; set; } = null;

        [JsonPropertyName("patientName")]
        [StringLength(60)]
        public string PatientName { get; set; } = "";

        [JsonPropertyName("patientShareAmount")]
        public double PatientShareAmount { get; set; } = 0;

        [JsonPropertyName("paymentMadeDate")]
        public DateTime? PaymentMadeDate { get; set; } = null;

        [JsonPropertyName("paymentReferenceNumber")]
        [StringLength(80)]
        public string PaymentReferenceNumber { get; set; } = "";

        [JsonPropertyName("payorFinalOfferAmount")]
        public double PayorFinalOfferAmount { get; set; } = 0;

        [JsonPropertyName("payorResolutionRequestReceivedDate")]
        public DateTime? PayorResolutionRequestReceivedDate { get; set; } = null;

        [JsonPropertyName("planPaidAmount")]
        public double PlanPaidAmount { get; set; } = 0;

        [JsonPropertyName("policyType")]
        [StringLength(40)]
        public string PolicyType { get; set; } = "";

        [JsonPropertyName("providerFinalOfferAmount")]
        public double ProviderFinalOfferAmount { get; set; } = 0;

        [JsonPropertyName("providerName")]
        [StringLength(60)]
        public string ProviderName { get; set; } = "";

        [JsonPropertyName("providerNPI")]
        [StringLength(10)]
        public string ProviderNPI { get; set; } = "";

        [JsonPropertyName("providerPaidDate")]
        public DateTime? ProviderPaidDate { get; set; } = null;

        [JsonPropertyName("providerType")]
        [StringLength(50)]
        public string ProviderType { get; set; } = "";

        [JsonPropertyName("reasonableAmount")]
        public double ReasonableAmount { get; set; } = 0;

        [JsonPropertyName("requestDate")]
        public DateTime? RequestDate { get; set; } = null;

        [JsonPropertyName("requestId")]
        [StringLength(20)]
        public string RequestId { get; set; } = "";

        [JsonPropertyName("requestType")]
        [StringLength(20)]
        public string RequestType { get; set; } = "";

        [JsonPropertyName("resolutionDeadlineDate")]
        public DateTime? ResolutionDeadlineDate { get; set; } = null;

        [JsonPropertyName("serviceDate")]
        public DateTime? ServiceDate { get; set; } = null;

        [JsonPropertyName("status")]
        [StringLength(50)]
        public string Status { get; set; } = "";

        [JsonPropertyName("submittedBy")]
        [StringLength(60)]
        public string SubmittedBy { get; set; } = "";

        [JsonPropertyName("totalSettlementAmount")]
        public double TotalSettlementAmount { get; set; } = 0;

        [JsonPropertyName("userId")]
        [StringLength(40)]
        public string UserId { get; set; } = "";

        [JsonPropertyName("wasIneligibilityDenied")]
        public bool WasIneligibilityDenied { get; set; } = false;

        [JsonPropertyName("wasDisputeSettledOutsideOfArbitration")]
        public bool WasDisputeSettledOutsideOfArbitration { get; set; } = false;

        [JsonPropertyName("wasDisputeSettledWithTeleconference")]
        public bool WasDisputeSettledWithTeleconference { get; set; } = false;

        [JsonPropertyName("wasSettledAtArbitration")]
        public bool WasSettledAtArbitration { get; set; } = false;

        [JsonPropertyName("wasPayorPaymentReceived")]
        public bool WasPayorPaymentReceived { get; set; } = false;

        [JsonPropertyName("wasPayorPaymentTimely")]
        public bool WasPayorPaymentTimely { get; set; } = false;

        [JsonPropertyName("wasProviderPaymentReceived")]
        public bool WasProviderPaymentReceived { get; set; } = false;

        [JsonPropertyName("wasProviderPaymentTimely")]
        public bool WasProviderPaymentTimely { get; set; } = false;

        [JsonPropertyName("winner")]
        [StringLength(50)]
        public string Winner { get; set; } = "";
    }
}

File: Arbitration/MPArbitration/ClientApp/src/app/model/tdi-request-details.ts
export class TDIRequestDetails
{
    public id = 0;
    public action = '';
    public additionalPaidAmount = 0;
    public arbitrationDate: Date|undefined;
    public arbitrationDecisionDate: Date|undefined;
    public arbitratorReportSubmissionDate: Date|undefined;
    public arbitrator1 = '';
    public arbitrator2 = '';
    public arbitrator3 = '';
    public arbitrator4 = '';
    public arbitrator5 = '';
    public assignmentDeadlineDate: Date|undefined;
    public batchUploadDate:Date|undefined;
    public payorClaimNumber = '';
    public createdBy = '';
    public daysOpen = 0;
    public disputedAmount = 0;
    public entityNPI = '';
    public entity = '';
    public estimatedDisputedAmount = 0;
    public healthPlanName = '';
    public history = '';
    public ineligibilityReason = '';
    public informalTeleconferenceDate:Date|undefined;
    public methodOfPayment = '';
    public originalBilledAmount = 0;
    public partiesAwardNotificationDate: Date|undefined;
    public patientName = '';
    public patientShareAmount = 0;
    public paymentMadeDate:Date|undefined;
    public paymentReferenceNumber = '';
    public payorFinalOfferAmount = 0;
    public payorResolutionRequestReceivedDate:Date|undefined;
    public planPaidAmount = 0;
    public policyType = '';
    public providerName = '';
    public providerFinalOfferAmount = 0;
    public providerNPI = '';
    public providerPaidDate: Date|undefined;
    public providerType = '';
    public reasonableAmount = 0;
    public requestDate:Date|undefined;
    public requestId = 0;
    public requestType = '';
    public resolutionDeadlineDate:Date|undefined;
    public serviceDate:Date|undefined;
    public status = '';
    public submittedBy = '';
    public totalSettlementAmount = 0;
    public userId = '';
    public wasIneligibilityDenied:Date|undefined;
    public wasDisputeSettledOutsideOfArbitration = false;        
    public wasDisputeSettledWithTeleconference = false;

    constructor(obj?:any) {
        if(!obj)
            return;
        Object.assign(this, JSON.parse(JSON.stringify(obj)));
        try {
            if(this.arbitrationDate)
                this.arbitrationDate = new Date(obj.arbitrationDate); 

            if(this.arbitrationDecisionDate)
                this.arbitrationDecisionDate = new Date(obj.arbitrationDecisionDate); 

            if(this.assignmentDeadlineDate)
                this.assignmentDeadlineDate = new Date(obj.assignmentDeadlineDate);

            if(this.batchUploadDate)
                this.batchUploadDate = new Date(obj.batchUploadDate); 
            if(this.informalTeleconferenceDate)
                this.informalTeleconferenceDate = new Date(obj.informalTeleconferenceDate); 
            if(this.partiesAwardNotificationDate)
                this.partiesAwardNotificationDate = new Date(obj.partiesAwardNotificationDate); 
            if(this.paymentMadeDate)
                this.paymentMadeDate = new Date(obj.paymentMadeDate); 
            if(this.payorResolutionRequestReceivedDate)
                this.payorResolutionRequestReceivedDate = new Date(obj.payorResolutionRequestReceivedDate);
            if(this.providerPaidDate)
                this.providerPaidDate = new Date(obj.providerPaidDate); 
            if(this.resolutionDeadlineDate)
                this.resolutionDeadlineDate = new Date(obj.resolutionDeadlineDate); 
            if(this.serviceDate)
                this.serviceDate = new Date(obj.serviceDate);
            if(this.wasIneligibilityDenied)
                this.wasIneligibilityDenied = new Date(obj.wasIneligibilityDenied); 
        }
        catch(err) {
            console.error('Error constructing TDIRequestDetails:');
            console.error(err);
        }
    }
}
